Added TimeSeriesFilteredController for filtered time series API, supp…

rma-rripken opened the pull request:
…orting query parameters like value range, null filtering, and pagination.

rma-rripken committed:
Added TimeSeriesFilteredController for filtered time series API, supporting query parameters like value range, null filtering, and pagination.

Clearing pagination so that it doesn't serialize twice
Pass the cursor timestamp into the retrieve_ts_entry_out_tab call when not using trim.

rma-rripken committed:
Added TimeSeriesFilteredController for filtered time series API, supporting query parameters like value range, null filtering, and pagination.

rma-rripken committed:
Added TimeSeriesFilteredController for filtered time series API, supporting query parameters like value range, null filtering, and pagination.

Clearing pagination so that it doesn't serialize twice
Pass the cursor timestamp into the retrieve_ts_entry_out_tab call when not using trim.

rma-rripken committed:
Added TimeSeriesFilteredController for filtered time series API, supporting query parameters like value range, null filtering, and pagination.

rma-rripken committed:
Added TimeSeriesFilteredController for filtered time series API, supporting query parameters like value range, null filtering, and pagination.

Clearing pagination so that it doesn't serialize twice
Pass the cursor timestamp into the retrieve_ts_entry_out_tab call when not using trim.

rma-rripken committed:
Added TimeSeriesFilteredController for filtered time series API, supporting query parameters like value range, null filtering, and pagination.

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesFilteredController.java:
```diff
@@ -0,0 +1,302 @@
+                            + "whether to include the data entry date of each value in the response. Including the data entry "
+                            + "date will increase the size of the array containing each data value from three to four, "
+                            + "changing the format of the response. Default is false."),
+                    @OpenApiParam(name = "asc", type = Boolean.class, description = "Specifies "
```
Would it make sense to add the "asc" parameter to the normal TimeSeries end-point?   It doesn't seem to be a filter and more about presentation.

MikeNeilson replied:
I feel like if the call *actually* wants the data in reverse order, they should just deal with it client side.

rma-rripken replied:
Should I remove the asc/desc flag from the Filtered end-point?  It was in the original request but I'm not sure how important that specific feature was.

MikeNeilson replied:
I say remove it. But give @krowvin an hour or two to chime in.

rma-rripken commented on cwms-data-api/src/main/java/cwms/cda/data/dao/TimeSeriesDao.java:
```diff
@@ -25,10 +26,31 @@ void store(TimeSeries timeSeries, boolean createAsLrts,
+    @Deprecated
     TimeSeries getTimeseries(String cursor, int pageSize, String names, String office,
                              String unit, ZonedDateTime begin, ZonedDateTime end,
                              ZonedDateTime versionDate, boolean trim, boolean includeEntryDate);
```
The number of parameters on this dao method are getting out of hand.

MikeNeilson replied:
Yep, though I think what you've done is probably the only practical solution. At least it groups things.

rma-rripken committed:
Removing broken mock test.

Introducing the TimeSeriesRequestParameter object broke the mock but we never really adopted that style of test.  Removing rather than trying to fix

rma-rripken committed:
Pass the cursor timestamp into the retrieve_ts_entry_out_tab call.

rma-rripken committed:
Pass the cursor timestamp into the retrieve_ts_entry_out_tab call.

rma-rripken committed:
Fixed retVal issue I inadvertently added.

rma-rripken committed:
Fixed test data reference

rma-rripken committed:
Can't change the start date for retrieve_ts_entry_out_tab when trim is being used.

rma-rripken committed:
Can't change the start date for retrieve_ts_entry_out_tab when trim is being used.

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesFilteredController.java:
```diff
@@ -0,0 +1,302 @@
+            boolean filterNulls = ctx.queryParamAsClass("filter-nulls", Boolean.class).getOrDefault(false);
+
+            // Having all the parameters in a RequestParameters class will make changing/extending the args easier.
+            FilteredTimeSeriesParameters ftsParams = new FilteredTimeSeriesParameters.Builder()
```
it seems like it would make sense to have a `fromContext` method in these builders. Not that they shouldn't have the individual methods, but that it does make sense for a "...Parameters" to know about the `Context:queryParam...` functions.

rma-rripken replied:
That would let FilteredTimeSeriesParameters find all the parameters it normally takes from the context by itself and the Controller wouldn't have to do it?   That could work.  Would clean up the controller code.

rma-rripken replied:
Tried this for FilteredTimeSeriesParameters.

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/api/TimeSeriesFilteredController.java:
```diff
@@ -0,0 +1,302 @@
+            // Send back the link to the next page in the response header
+            StringBuilder linkValue = new StringBuilder(600);
+            String pageUrl = buildRequestUrl(ctx, fts.getPage());
+            linkValue.append(String.format("<%s>; rel=self; type=\"%s\"",
```
why aren't these just embedded in the opaque next-page cursor value?
That was the point of handling the cursor in the way we do, to be able to encode such information without the user needing to worry about those particulars.

rma-rripken replied:
Does anyone or anything use the Link header value?  I was only doing it b/c the TimeSeriesController was doing it and that has been doing it from the start.

rma-rripken replied:
Some of the page and next-page tokens encode more information than others.  I don't think any of them include every query parameter - they mostly include the cursor part.  Its mostly been up to the caller to take the token and add it as a parameter to the url.   At least that is how I understand it.   I think this Link header is only done on these two controller but its kinda nice in that is builds the full url.

MikeNeilson replied:
it's supposed to be a single thing to said to the system so it *should* encode every parameter relevant to retrieving the next page of results.

I can see the value in both for some usages. but the page/cursor should be "all of it."

MikeNeilson replied:
As far as the link header, I think only the time series. and I think @DanielTOsborne set that up original. I think it was before we settled on the cursor as well.

rma-rripken replied:
A downside of the page containing every parameter is that we have to remember whenever we add Controller arguments that we also have to change the page parsing code.  If we switch to using Parameter objects there might be a better pattern.   Like if the user passed "page" we call fromPage(str) to build the parameter and if they didn't we call fromContext(context) to build it.

MikeNeilson replied:
hmm. I think I like that idea. but it's definitely not the way we've, or at least I've, been thinking about and doing it. So a bit beyond the scope of this PR perhaps.

Given that this particular end point was already a bit of a special case, I think we leave what you have for know and we do a little investigation on the paging design after.

rma-rripken replied:
Ok, I like the idea of the Controller or even the Parameter object parsing the paging cursor into a new Parameter object and the Dao not having to know anything about the paging token.  Might have to draw some pictures or have a call to make this really clear.  I think we can come up with a better pattern

MikeNeilson replied:
Oh, that makes sense. the DAO just handle actual values query values and some other component is responsible for the actual interpretation. Would allow either an opaque cursor or a navlink setup to be used with minimal difference, none of which is in the DAO itself.

MikeNeilson reviewed (commented):
Looks reasonable to me so far.

rma-rripken committed:
Clearing pagination so that it doesn't serialize twice

rma-rripken committed:
Clearing pagination so that it doesn't serialize twice

rma-rripken committed:
Moved new filter controller constants into Controllers.

rma-rripken committed:
Moved new filter controller constants into Controllers.

rma-rripken committed:
Moved new filter controller constants into Controllers.

rma-rripken committed:
Moved new filter controller constants into Controllers.

rma-rripken committed:
Added date checking to filtered IT

rma-rripken committed:
Added date checking to filtered IT

rma-rripken committed:
Merge remote-tracking branch 'origin/feature/2088-ts-filter' into feature/2088-ts-filter

zack-rma committed:
Fixes 1017: TimeSeries retrieval unit conversion error handling (#1119)

Fixes #1017

Adds global exception handling for unit conversion errors to JOOQ DAO
base class. Adds tests for timeseries, levels, and basins.

Updates documentation for Ratings and Locations, as controllers were
mislabeled to allow units beyond just unit systems as input.

Swaps string query parameters in test for constants.

MikeNeilson committed:
Feature/1092 1098 basic user management (#1115)

I think there's still a bit more to do, but ready for discussion. Plus
the tests just started failing on my laptop after adding one more (like
fail to even start correctly) so checking on github.

closes #1092 
closes #1098

zack-rma committed:
Fixes for LocationLevel Deserialization (#1159)

Fixed TimeSeriesLocationLevel deserialization in CREATE endpoint. Added
storage tests for TimeSeriesLocationLevel, SeasonalLocationLevel,
ConstantLocationLevel. Updated handling of interpolate string in storage
method of DAO. Set default values interval month and minutes for
SeasonalValueBean - resolves a bug in the storage procedure that
requires interval month and minutes to be non-null. Adds support for
total results to catalog endpoint.

rma-bryson committed:
CWMS-2096 - Updates ratings to allow for replacing base curve.  (#1160)

Fixes https://github.com/USACE/cwms-data-api/issues/624

Fixes issue with errors not propagating. Fixes issue with wrong query
parameter being added in for replace base curve. Adds logic to remove
template if store-template flag is false.

rma-rripken committed:
Remove ascending parameter from FilteredTimeSeriesTest and TimeSeriesDaoImpl

rma-rripken committed:
Remove ascending parameter from FilteredTimeSeriesTest and TimeSeriesDaoImpl

rma-rripken committed:
Feature/2088 rsql (#1164)

rma-rripken committed:
Feature/2088 rsql (#1164)

rma-rripken committed:
Feature/2088 rsql (#1164)

rma-rripken committed:
Feature/2088 rsql (#1164)

rma-rripken committed:
Merge remote-tracking branch 'origin/feature/2088-ts-filter' into feature/2088-ts-filter

MikeNeilson commented on cwms-data-api/src/main/java/cwms/cda/ApiServlet.java:
```diff
@@ -676,6 +688,20 @@ protected void configureRoutes() {
         addProjectLockRightsHandlers("/project-lock-rights/{project-id}", requiredRoles);
+
+
+        addUserManagementHandlers();
```
I think you're source branch needs updating. these are already in the target branch

rma-rripken replied:
Yes.  thank you for catching this, I will revisit.  I think the user management was merged after my branch started.  I did rebase but maybe this was missed.

MikeNeilson commented on cwms-data-api/src/main/webapp/rsql.html:
```diff
@@ -0,0 +1,118 @@
+<!DOCTYPE html>
```
I'm not going to worry about it at the moment, and frankly this will be useful anyways, but we do now have the Read-The-Docs documentation setup in the docs folder.

MikeNeilson commented on cwms-data-api/src/main/webapp/times.html:
```diff
@@ -0,0 +1,122 @@
+<!DOCTYPE html>
```
this one probably should've gone in the docs folder. But again, better to have something now, we'll worry about it later.

MikeNeilson commented on cwms-data-api/src/test/java/cwms/cda/data/dao/rsql/RSQLConditionBuilderTest.java:
```diff
@@ -0,0 +1,187 @@
+        assertNotNull(condition);
+        String conditionString = condition.toString();
+
+        assertTrue(conditionString.contains("\"AV_TSV_DQU\".\"UNIT_ID\" = 'EN'"));
```
do these end up as bind variables by the time it gets to the database?

MikeNeilson replied:
Based on the visitor above it seems like they should, so I'm surprised you can test this way, unless it's just part of the `Condition::toString`?

rma-rripken replied:
This test is relying on jOOQs Condition::toString.  Its not actually talking to the database, just using the view that it knows about from the jOOQ codegen.   I have two "Resolvers" that take the RSQL selectors and connect them to jOOQ fields.   I was hoping the JooqFieldResolver used in this test was going to make the timeseries impl really nice and clean.  It works when the jOOQ "from" clause is a simple view.  In reality the timeseries query is more complicated and I ended up using a different Resolver in the actual Dao code.  The JooqFieldResolver does work and makes the unit test cleaner.

rma-rripken replied:
This example is operating on AV_TSV_DQU.UNIT_ID from the view.    JooqFieldResolver can find unit_id b/c it looks at all the table/view fields.    The Resolver I used in the actual Dao impl only knows about the fields you tell it about and I didn't include UNIT_ID b/c I figured it didn't make real-world sense.  Why would people give the end-point a timeseries_id and then ask it to filter by unit_id.  Didn't make real-world sense to me.   But RSQLConditionBuilder is written to potentially be more generic than just the timeseries usage so I just left the test in there.  It demonstrates that the RSQL syntax can have String comparisons.

MikeNeilson reviewed (commented):
I have one paranoid question relating to the generated queries.

But otherwise this look great.

Though something is wonky with the pull request, when looking at the full commits it's claiming to add the new usermanagement things we merged in recently, not sure what's going on there.

zack-rma committed:
Issue 1107: Add rate limiting to rate endpoints (#1161)

Fixes #1107

Adds rate limiting to ratings endpoints. Designed for future generalized
application to unauthorized endpoints.

rma-bryson committed:
Issue #1113 - Updates to use 'units' as a query parameter to match timeseries json payload 'units'. (#1165)

Fixes https://github.com/USACE/cwms-data-api/issues/1113

rohaan-rma committed:
Add Rohaan to CONTRIBUTORS.md (#1166)

Adding myself as a contributor

Co-authored-by: Rohaan Sandhu <rohsan4845@geiconsultants.com>

zack-rma committed:
Issue 1080: Location Post Update (#1163)

Fixes #1080

Adds `FAIL-IF-EXISTS` query parameter to locations create endpoint. Flag
defaults to true.
